pkg/controller/daemon/util/daemonset_util.go

```diff
@@ -177,22 +176,25 @@ func ReplaceDaemonSetPodHostnameNodeAffinity(affinity *v1.Affinity, nodename str
 
 	nodeSelectorTerms := []v1.NodeSelectorTerm{}
 
-	// Removes hostname node selector, as only the target hostname will take effect.
+	// Removes nodename node selector, as only the target nodename will take effect.
 	for _, term := range nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
-		exps := []v1.NodeSelectorRequirement{}
-		for _, exp := range term.MatchExpressions {
-			if exp.Key != kubeletapis.LabelHostname {
-				exps = append(exps, exp)
+		// Removes nodename from term.MatchFields if not empty
+		if len(term.MatchFields) != 0 {
+			exps := []v1.NodeSelectorRequirement{}
+			for _, exp := range term.MatchFields {
+				if exp.Key != algorithm.NodeFieldSelectorKeyNodeName {
+					exps = append(exps, exp)
+				}
 			}
+			term.MatchFields = exps
 		}
 
-		if len(exps) > 0 {
-			term.MatchExpressions = exps
+		if len(term.MatchFields) != 0 || len(term.MatchExpressions) != 0 {
```

Comment:
This could break if we add another field to `NodeSelectorTerm` in the future. That new field could removed here if the other two fields are both empty.

Reply:
Yes, that's true; it seems we did not whether other fields in term, otherwise, we leave a empty term here to cover that.

Reply:
> Why don't we append the term to nodeSelectorTerms unconditionally?

As `nil or empty` requirements select no objects. 


Did an update to append target node affinity to all terms: it's ORed between terms, and ANDed with a term; and there's no conflict as this target is subset of the node selector.